File: src/main/java/name/wind/tools/eq2/lp/EventBrowserStageController.java
package name.wind.tools.eq2.lp;

import javafx.event.ActionEvent;
import javafx.fxml.FXML;
import javafx.scene.control.TreeTableView;
import javafx.stage.FileChooser;
import javafx.stage.Modality;
import javafx.stage.Stage;
import name.wind.application.cdi.fx.StageController;
import name.wind.application.cdi.fx.annotation.FXMLResource;
import name.wind.application.cdi.fx.event.FXMLResourceOpen;
import name.wind.common.util.KnownSystemProperties;
import name.wind.common.util.Pipeliner;
import name.wind.tools.eq2.lp.log.LogRecord;

import javax.enterprise.context.ApplicationScoped;
import javax.enterprise.event.Event;
import javax.inject.Inject;
import java.io.File;
import java.util.HashMap;
import java.util.Optional;

@ApplicationScoped @FXMLResource(Globals.FXMLResources.FXML__EVENT_BROWSER_STAGE) public class EventBrowserStageController
    extends StageController implements KnownSystemProperties {

    @Inject private Event<FXMLResourceOpen> fxmlResourceOpenEvent;
    @FXML private TreeTableView<LogRecord> eventTreeTableView;

    @FXML protected void openFile(ActionEvent event) {
        Optional.ofNullable(Pipeliner.of(FileChooser::new)
            .set(target -> target::setInitialDirectory, Optional.ofNullable(Globals.Settings.openFileInitialDirectory.get())
                .orElse(userHomeFile.get().orElse(
                    new File(""))))
            .map(fileChooser -> fileChooser.showOpenMultipleDialog(stage))
            .get()).ifPresent(files -> {
                files.stream().findAny().map(File::getParentFile).ifPresent(Globals.Settings.openFileInitialDirectory);
                fxmlResourceOpenEvent.fire(
                    new FXMLResourceOpen(
                        Pipeliner.of(Stage::new)
                            .set(target -> target::initOwner, stage)
                            .set(target -> target::initModality, Modality.APPLICATION_MODAL)
                            .set(target -> target::setResizable, false)
                            .get(),
                        Globals.FXMLResources.FXML__SELECT_LOG_EVENT_BUILDER_STAGE,
                        Pipeliner.of(HashMap<String, Object>::new)
                            .accept(target -> target.put("files", files))
                            .get()));
            });
    }

}

File: src/main/kotlin/com/github/windchopper/tools/everquest/log/parser/ui/EventBrowserStageController.kt
@file:Suppress("NestedLambdaShadowedImplicitParameter")

package com.github.windchopper.tools.everquest.log.parser.ui

import com.github.windchopper.common.fx.cdi.form.Form
import com.github.windchopper.common.fx.cdi.form.StageFormController
import com.github.windchopper.common.fx.cdi.form.StageFormLoad
import com.github.windchopper.common.util.ClassPathResource
import com.github.windchopper.tools.everquest.log.parser.Application
import jakarta.enterprise.context.ApplicationScoped
import jakarta.enterprise.event.Event
import jakarta.inject.Inject
import javafx.event.ActionEvent
import javafx.fxml.FXML
import javafx.stage.FileChooser
import javafx.stage.Modality
import javafx.stage.Stage
import java.io.File

@ApplicationScoped @Form(Application.FXML__EVENT_BROWSER_STAGE) class EventBrowserStageController: StageFormController() {

    @Inject private lateinit var stageFormLoadEvent: Event<StageFormLoad>

    @FXML fun openFile(event: ActionEvent) {
        val files = FileChooser().let {
            it.initialDirectory = Application.openFileInitialDirectory.load().value?:File(System.getProperty("user.home"))
            it.showOpenMultipleDialog(stage)
        }

        files?.let { nonEmptyFiles ->
            Application.openFileInitialDirectory.save(nonEmptyFiles.first().parentFile)
            stageFormLoadEvent.fire(StageFormLoad(
                ClassPathResource(Application.FXML__LOG_FILE_SELECTION_STAGE),
                mapOf("files" to nonEmptyFiles)) {
                    Stage().also { newStage ->
                        newStage.initOwner(stage)
                        newStage.initModality(Modality.APPLICATION_MODAL)
                        newStage.isResizable = false
                        newStage.title = Application.messages["logFileSelectionStage.title"]
                    }
                })
        }
    }

}